dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/AbstractEventService.java

```diff
@@ -1933,14 +2020,70 @@ private OrganisationUnit getOrganisationUnit( IdSchemes idSchemes, String id )
             () -> manager.getObject( OrganisationUnit.class, idSchemes.getOrgUnitIdScheme(), id ) );
     }
 
-    private Program getProgram( IdScheme idScheme, String id )
+    private ProgramStageInstance getProgramStageInstance( String uid )
     {
-        return programCache.get( id, () -> manager.getObject( Program.class, idScheme, id ) );
+        ProgramStageInstance programStageInstance = programStageInstanceCache.get( uid );
+        
+        if ( programStageInstance == null )
+        {
+            programStageInstance = eventSyncService.getEvent( uid );
+            
+            programStageInstanceCache.put( uid, programStageInstance );
```

Comment:
yes, have thought about this as well... just wanted to have few lines :(

Reply:
As I said. If it was an intention and you know about it. Then perfectly fine. Maybe adding 1 line of comment with explanation would be fine. But sure, I understand that it will work and it is fine (if explained so other people reading it know that it is not a bug).